crates/chain/src/local_chain.rs

```diff
@@ -19,215 +170,250 @@ impl ChainOracle for LocalChain {
     fn is_block_in_chain(
         &self,
         block: BlockId,
-        static_block: BlockId,
+        chain_tip: BlockId,
     ) -> Result<Option<bool>, Self::Error> {
-        if block.height > static_block.height {
+        if block.height > chain_tip.height {
             return Ok(None);
         }
         Ok(
             match (
-                self.blocks.get(&block.height),
-                self.blocks.get(&static_block.height),
+                self.index.get(&block.height),
+                self.index.get(&chain_tip.height),
             ) {
-                (Some(&hash), Some(&static_hash)) => {
-                    Some(hash == block.hash && static_hash == static_block.hash)
-                }
+                (Some(cp), Some(tip_cp)) => Some(*cp == block.hash && *tip_cp == chain_tip.hash),
                 _ => None,
             },
         )
     }
 
     fn get_chain_tip(&self) -> Result<Option<BlockId>, Self::Error> {
-        Ok(self.tip())
+        Ok(self.tip.as_ref().map(|tip| tip.block_id()))
     }
 }
 
-impl AsRef<BTreeMap<u32, BlockHash>> for LocalChain {
-    fn as_ref(&self) -> &BTreeMap<u32, BlockHash> {
-        &self.blocks
-    }
-}
+impl LocalChain {
+    /// Construct a [`LocalChain`] from an initial `changeset`.
+    pub fn from_changeset(changeset: ChangeSet) -> Self {
+        let mut chain = Self::default();
+        chain.apply_changeset(&changeset);
 
-impl From<LocalChain> for BTreeMap<u32, BlockHash> {
-    fn from(value: LocalChain) -> Self {
-        value.blocks
+        debug_assert!(chain._check_index_is_consistent_with_tip());
+        debug_assert!(chain._check_changeset_is_applied(&changeset));
+
+        chain
     }
-}
 
-impl From<BTreeMap<u32, BlockHash>> for LocalChain {
-    fn from(value: BTreeMap<u32, BlockHash>) -> Self {
-        Self { blocks: value }
+    /// Construct a [`LocalChain`] from a given `checkpoint` tip.
+    pub fn from_tip(tip: CheckPoint) -> Self {
+        let mut _self = Self {
```

Comment:
nit: this should be `self` as it is used below

Reply:
I don't think that's legal. We might get arrested by the rustacean police.